packages/sitecore-jss-react/src/components/FEaaSComponent.tsx

```diff
@@ -1,94 +1,119 @@
-    !params.ComponentRevision ||
-    !params.ComponentHostName
+    (!props.lastModified || !props.template) &&
+    !props.src &&
+    (!props.params ||
+      !props.params.LibraryId ||
+      !props.params.ComponentId ||
+      !props.params.ComponentVersion ||
+      !props.params.ComponentRevision ||
+      !props.params.ComponentHostName)
   ) {
     // Missing FEaaS component required props
     return null;
   }
 
-  // Load script asynchronously on the page once, only if FEAAS component is used
-  useEffect(() => {
-    window.FEAASLoading ??= new Promise((resolve) => {
-      const script = document.createElement('script');
-      script.src = FEAAS_MODULE_SRC;
-      script.type = 'module';
-      script.onload = resolve;
-      document.head.appendChild(script);
-    });
-  }, []);
-
-  let data = '';
-  if (params?.ComponentDataOverride) {
+  let data: string = null;
+  if (props.params?.ComponentDataOverride) {
     // Use override data if provided
-    data = params.ComponentDataOverride;
-  } else if (fields) {
+    data = props.params.ComponentDataOverride;
+  } else if (props.fields) {
     // Otherwise use datasource data (provided in fields)
     // FEaaS expects wrapping "_" as catch-all datasource template id
-    data = JSON.stringify({ _: getDataFromFields(fields) });
-  }
-
-  const reqProps = {
-    library: params.LibraryId,
-    cdn: params.ComponentHostName,
-    component: params.ComponentId,
-    version: params.ComponentVersion,
-    revision: params.ComponentRevision,
-    data,
-  };
-  const optProps: { [key: string]: unknown } = {};
-  if (params.ComponentInstanceId) {
-    optProps.instance = params.ComponentInstanceId;
-  }
-  if (params.ComponentHTMLOverride) {
-    optProps.dangerouslySetInnerHTML = { __html: params.ComponentHTMLOverride };
+    data = JSON.stringify({ _: getDataFromFields(props.fields) });
   }
 
+  // FEaaS control would still be hydrated by client
+  // we pass all the props as a workaround to avoid hydration error, until we convert all JSS components to server side
+  // this also allows component to fall back to full client-side rendering when template or src is empty
   return (
-    <>
-      <link rel="preload" as="style" href={`${reqProps.cdn}/styles/${reqProps.library}/published.css`} />
-      <link rel="preload" as="fetch" href={`${reqProps.cdn}/components/${reqProps.library}/${reqProps.component}/${reqProps.version}/${reqProps.revision}.html`} />
-      <link rel="preload" as="script" href={FEAAS_MODULE_SRC} />
-      <feaas-stylesheet library={reqProps.library} cdn={reqProps.cdn} />
-      <feaas-component {...reqProps} {...optProps} />
-    </>
+    <FEAAS.Component
+      data={data}
+      template={props.template}
+      last-modified={props.lastModified}
+      src={props.src}
+      cdn={props.params?.ComponentHostName}
+      library={props.params?.LibraryId}
+      component={props.params?.ComponentId}
+      revision={props.params?.ComponentRevision}
```

Comment:
Do we need to add version here? `props.params.ComponentVersion` is unused atm.

Reply:
added